pkg/apis/storage/types.go

```diff
@@ -215,3 +215,174 @@ const (
+	// method), and that Kubernetes should call attach and wait for any attach
+	// operation to complete before proceeding to mounting.
+	// If value is not specified, default is false -- meaning attach will not be
+	// called.
+	// +optional
+	AttachRequired *bool
+
+	// If specified, podInfoRequiredOnMount indicates this CSI volume driver
+	// requires additional pod information (like podName, podUID, etc.) during
+	// mount operations.
+	// If value is not specified, pod information will not be passed on mount.
+	// If value is set to a valid version, Kubelet will pass pod information as
+	// VolumeAttributes in the CSI NodePublishVolume() calls.
+	// Supported versions:
+	// Version "v1" will pass the following ValueAttributes
+	// "csi.storage.k8s.io/pod.name": pod.Name
+	// "csi.storage.k8s.io/pod.namespace": pod.Namespace
+	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
+	// +optional
+	PodInfoOnMountVersion *string
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSINodeInfo holds information about all CSI drivers installed on a node.
+type CSINodeInfo struct {
+	metav1.TypeMeta
+
+	// metadata.name must be the Kubernetes node name.
+	metav1.ObjectMeta
+
+	// spec is the specification of CSINodeInfo
+	Spec CSINodeInfoSpec
+
+	// status is the current status of CSINodeInfo
+	Status CSINodeInfoStatus
+}
+
+// CSINodeInfoSpec holds information about the specification of all CSI drivers installed on a node
+type CSINodeInfoSpec struct {
+	// drivers is a list of specifications of CSIDriverInfo
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoSpec
+}
+
+// CSINodeInfoStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeInfoStatus struct {
+	// drivers is a list of the statuses of CSIDriverInfo
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoStatus
+}
+
+// CSIDriverInfoStatus holds information about the status of one CSI driver installed on a node
+type CSIDriverInfoStatus struct {
+	// name is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string
+
+	// available is a boolean representing whether the driver has been installed
+	// on this node or not.
+	Available bool
+
+	// volumePluginMechanism announces what mechanism underlies volume plugins.
+	// It is set by Kubelet. It is used by the attach/detach controller, which
+	// needs to know how to perform attachments. The allowed values are:
+	// * "in-tree": the volume operation (e.g., attach/detach) ought to be
+	//   directly performed by the attach/detach controller.
+	// * "csi-plugin": the attach/detach controller ought to request
+	//   the csi plugin to perform the volume operation rather than perform it directly.
+	VolumePluginMechanism VolumePluginMechanism
+}
```

Comment:
Did we want to keep `VolumePluginMechanism` bit in a separate commit since this commit is more about just migrating these types in-tree and keeping same API?

Reply:
`VolumePluginMechanism` has already been a part of the CRD for a while now :)

Reply:
ah, disregard this. :-)

Reply:
Hm but this field is only used for migration which is still going to be alpha?

Reply:
@msau42 has a good point. Although its been in the CRD it hasn't really gotten any usage and we're not 100% sure that its exactly what we want. We may want to add an Alpha API with this field (any gating can be done with `CSIMigration`) so that we have some flexibility to change it in the future when we start using it. @xing-yang what do you think?

Reply:
Sure, I can make VolumePluginMechanism an Alpha field.  Can I keep it together in the same commit with other API changes or should it be in a separate commit in this PR?

Reply:
whichever works better for you is ok here

Reply:
Do we want to make it an alpha field in the beta api object, or have an entirely new alpha object?

I'm trying to think about the challenges of making it an entirely separate object.  It means all intree and external controller code needs to be able to handle both beta and alpha api groups.

Reply:
I’m looking into this too.

I see this PR that adds volume resize support as an Alpha feature: 
https://github.com/kubernetes/kubernetes/pull/49727/commits/e78d4331501d26c0fb4cb22514a6ba70296d7e34

The new PersistentVolumeClaimCondition struct was added to staging/src/k8s.io/api/core/v1/types.go, and it was added as a new field to PersistentVolumeClaimStatus.

Then there is a "ExpandPersistentVolumes" Alpha feature gate that is used for validation.

Can we do something similar here?  Keep this field VolumePluginMechanism in the Beta API object, but add validation for it using the feature gate.

Reply:
But setting of conditions field was feature gated(for resizing) and hence if you do not have feature gate enabled, you can not interact with the new field. It would appear to you that, field does not exist.

Reply:
Discussed this with @davidz627 .

We think a better name for this field could be `ShimMechanism` or `ShimMode` with values "InTree" or "CSI".  It should be a pointer so that CSI drivers with no in-tree equivalent can leave it blank.

Reply:
Sure, I'll rename this.  I think this is already changed to a pointer in the latest PR.

Reply:
or `PluginShimMode`

Reply:
Would `VolumePlugin` be redundant since we're already in a CSIDriver object?

Reply:
I'll change this to ShimMode then.

Reply:
As discussed [here](https://github.com/kubernetes/kubernetes/pull/73883#discussion_r258638861), we decided to remove this field (and the entire status field)

Reply:
Oh, that's right!  I'm going to remove the status field first so I won't get myself confused:).